pkg/workflow/step/generator.go

```diff
@@ -54,6 +57,27 @@ func NewChainWorkflowStepGenerator(generators ...WorkflowStepGenerator) Workflow
 	return &ChainWorkflowStepGenerator{generators: generators}
 }
 
+// RefWorkflowStepGenerator generate workflow steps from ref workflow
+type RefWorkflowStepGenerator struct {
+	context.Context
+	client.Client
+}
+
+// Generate generate workflow steps
+func (g *RefWorkflowStepGenerator) Generate(app *v1beta1.Application, existingSteps []v1beta1.WorkflowStep) (steps []v1beta1.WorkflowStep, err error) {
+	if len(existingSteps) > 0 {
+		return existingSteps, nil
+	}
+	if app.Spec.Workflow != nil && app.Spec.Workflow.Ref != "" {
```

Comment:
```suggestion
        if app.Spec.Workflow == nil || app.Spec.Workflow.Steps != nil {
             return existingSteps, nil
        } 
	if  app.Spec.Workflow.Ref == "" {
	    return errors.New("either workflow steps or ref should not be empty")
	}
```

you should not judge by if there's existingSteps, you should judge by if the app.Spec.Workflow is not nil.

Reply:
Not exactly. If ref and steps are both null, we will generate workflow steps according to Components and Policies.

Reply:
But we should add logic to avoid user setting workflow ref and steps at the same time here. I add it now.